### PR TITLE
Lex '''-delimited multiline string literals.

### PR DESCRIPTION
**Pull request by zygoloid**
"""-delimited literals are still lexed for error recovery but produce an error.

### CODE REVIEW

File: toolchain/lexer/string_literal.cpp [hunk 1]
```diff
@@ -53,12 +69,14 @@ auto LexedStringLiteral::Lex(llvm::StringRef source_text)
...
 
-  const int multi_line_prefix_size =
-      GetMultiLineStringLiteralPrefixSize(source_text.substr(hash_level));
-  const bool multi_line = multi_line_prefix_size > 0;
+  const int multi_line_prefix_size = GetMultiLineStringLiteralPrefixSize(
+      source_text.substr(hash_level), terminator);
+  const MultiLineKind multi_line = multi_line_prefix_size == 0 ? NotMultiLine
+                                   : terminator[0] == '"'
+                                       ? MultiLineWithDoubleQuotes
+                                       : MultiLine;
```
Comment (chandlerc) [inline toolchain/lexer/string_literal.cpp]: I feel like this is getting to be a bit hard to follow with the densely encoded information across the API boundary. What do you think about a simpler approach of returning a tuple of the kind and size?

File: toolchain/lexer/string_literal.cpp [hunk 2]
```diff
@@ -103,9 +121,11 @@ auto LexedStringLiteral::Lex(llvm::StringRef source_text)
...
         }
         break;
-      case '\"': {
-        if (terminator.size() == 1 ||
-            source_text.substr(cursor).startswith(terminator)) {
+      case '"':
+      case '\'': {
+        if (terminator.size() == 1
+                ? source_text[cursor] == '"'
+                : source_text.substr(cursor).startswith(terminator)) {
```
Comment (chandlerc) [inline toolchain/lexer/string_literal.cpp]: I find a ternary expression of conditions ... quite hard to read. Do you see any other way to write this? I'm OK if not, but seemed worth checking.

**Review by chandlerc — CHANGES_REQUESTED** [review]
Nice, mostly minor readability comments inline.

File: toolchain/lexer/string_literal.cpp (hunk 2 above)
Comment (jonmeow) [reply toolchain/lexer/string_literal.cpp:128]: Looking at this code, is `"'"` tested? I think it may not be:
terminator == "\""
source_text[cursor] == "'"

Looks like it may break incorrectly, suggesting this needs to be rewritten?

Comment (jonmeow) [reply toolchain/lexer/string_literal.cpp]: This was performance focused -- you might want to look at how the benchmark behaves. It may be that with the ternary, only calling `source_text.substr(cursor).startswith(terminator)` starts yielding similar results, but not sure.

File: toolchain/lexer/string_literal.cpp (hunk 1 above)
Comment (zygoloid) [reply toolchain/lexer/string_literal.cpp]: Thanks for the suggestion. I ended up with a named class rather than a tuple; I think this is substantially clearer than what I had before.

File: toolchain/lexer/string_literal.cpp (hunk 2 above)
Comment (zygoloid) [reply toolchain/lexer/string_literal.cpp]: Before:
```
-----------------------------------------------------------------------------
Benchmark                                   Time             CPU   Iterations
-----------------------------------------------------------------------------
BM_ValidString_Simple                  167867 ns       167853 ns         4171
BM_ValidString_Multiline               126114 ns       126107 ns         5554
BM_ValidString_Raw                     168004 ns       168002 ns         4167
BM_IncompleteWithEscapes_Simple        177253 ns       177253 ns         3952
BM_IncompleteWithEscapes_Multiline     167854 ns       167822 ns         4171
BM_IncompleteWithEscapes_Raw           344739 ns       344697 ns         2030
```
After:
```
----------------------------------------------------------------------------------------
Benchmark                                              Time             CPU   Iterations
----------------------------------------------------------------------------------------
BM_ValidString_Simple                             168197 ns       168194 ns         4169
BM_ValidString_Multiline                          168073 ns       168069 ns         4173
BM_ValidString_MultilineDoubleQuote               167996 ns       167990 ns         4169
BM_ValidString_Raw                                167826 ns       167817 ns         4166
BM_IncompleteWithEscapes_Simple                   188479 ns       188460 ns         3705
BM_IncompleteWithEscapes_Multiline                167887 ns       167883 ns         4165
BM_IncompleteWithEscapes_MultilineDoubleQuote     167936 ns       167928 ns         4170
BM_IncompleteWithEscapes_Raw                      372160 ns       372160 ns         1895
```

Removing the special case entirely makes the BM_IncompleteWithEscapes benchmarks a fair bit slower:
```
----------------------------------------------------------------------------------------
Benchmark                                              Time             CPU   Iterations
----------------------------------------------------------------------------------------
BM_ValidString_Simple                             168077 ns       168072 ns         4160
BM_ValidString_Multiline                          168088 ns       168079 ns         4171
BM_ValidString_MultilineDoubleQuote               167963 ns       167960 ns         4173
BM_ValidString_Raw                                168113 ns       168109 ns         4173
BM_IncompleteWithEscapes_Simple                   287398 ns       287368 ns         2440
BM_IncompleteWithEscapes_Multiline                280911 ns       280881 ns         2494
BM_IncompleteWithEscapes_MultilineDoubleQuote     280847 ns       280839 ns         2491
BM_IncompleteWithEscapes_Raw                      389369 ns       389369 ns         1802
```
... which doesn't superficially make any sense because those benchmarks never even hit the switch cases that change. This is presumably a consequence of the optimizer making a different decision about how to structure the loop.

Similarly, it looks like the difference for ValidString_Multiline is all about whether LLVM happens to unswitch the loop, and I've not found any reliable way to "encourage" that to happen except by forcing it in the source code.

Overall, I think my attempts at optimizing further here have just been chasing whatever perturbation of the source code happens to cause the optimizer to emit a faster loop today, and it's not worth spending more time here.

Comment (zygoloid) [reply toolchain/lexer/string_literal.cpp]: To @chandlerc's comment: I've restructured slightly to avoid the ternary. Not super happy about the result (yuck, a `goto`!), but avoiding the length-one check when we hit a `'` does seem to make benchmarks run a bit faster.

Comment (zygoloid) [reply toolchain/lexer/string_literal.cpp]: It seems that nearly all the performance deltas we were seeing were due to how LLVM was choosing to lower the `switch` into a sequence of comparisons. In most of the benchmarks, we weren't spending almost any time in the actual case bodies, but rather were spending most of the time figuring out if we wanted to enter a case at all.

I've added a separate lookup table to check whether a character is special before entering the switch, which seems to help a lot. Performance of the revised approach:
```
----------------------------------------------------------------------------------------
Benchmark                                              Time             CPU   Iterations
----------------------------------------------------------------------------------------
BM_ValidString_Simple                             134646 ns       134640 ns         5204
BM_ValidString_Multiline                          134705 ns       134694 ns         5202
BM_ValidString_MultilineDoubleQuote               135015 ns       135009 ns         5103
BM_ValidString_Raw                                140113 ns       140087 ns         4838
BM_IncompleteWithEscapes_Simple                   148642 ns       148626 ns         4672
BM_IncompleteWithEscapes_Multiline                140090 ns       140074 ns         4867
BM_IncompleteWithEscapes_MultilineDoubleQuote     139795 ns       139788 ns         4901
BM_IncompleteWithEscapes_Raw                      321821 ns       321817 ns         2191
BM_SimpleStringValue_Simple                       323073 ns       323051 ns         2371
BM_SimpleStringValue_Multiline                    293697 ns       293675 ns         2390
BM_SimpleStringValue_MultilineDoubleQuote         306823 ns       306791 ns         2391
BM_SimpleStringValue_Raw                          320943 ns       320904 ns         2360
```
This is basically tied with the best other approach @chandlerc and I had been looking at, wherein we generated specialized versions of the loop for each kind of string and for each number of hashes up to 2 -- except for the WithEscapes_Raw benchmark, where we're still seeing some cost from repeated string comparisons against the hashes string, but I'm somewhat less concerned by the performance for raw string literals, and in any case this beats what we had before.

With this change, the point optimization to check the length-one case for the terminator actually makes performance worse, so I've removed it. The corresponding optimization for escape sequences is still worthwhile, though. Performance here is still quite fragile, for example adding a `default: llvm_unreachable(...);` to the `switch` makes some of the benchmarks 50% slower. But given that the current benchmarks are clocking in at ~3 cycles per source byte, this level of fragility is perhaps not surprising.

Comment (chandlerc) [reply toolchain/lexer/string_literal.cpp]: Yeah, I think we've nerd sniped ourselves sufficiently for now. ;]

**Review by chandlerc — APPROVED** [review]
LG
